tree/ntuple/v7/inc/ROOT/RNTupleCollectionWriter.hxx

```diff
@@ -30,43 +33,65 @@ namespace Experimental {
 /**
 \class ROOT::Experimental::RNTupleCollectionWriter
 \ingroup NTuple
-\brief A virtual ntuple used for writing untyped collections that can be used to some extent like an RNTupleWriter
-*
-* This class is between a field and a ntuple.  It carries the offset column for the collection and the default entry
-* taken from the collection model.  It does not, however, own an ntuple model because the collection model has been
-* merged into the larger ntuple model.
+\brief A special type only prodcued by the collection field and used for writing untyped collections
+
+This class is tightly coupled to the RCollectionField. It fills the sub fields of the collection fields one-by-one.
+An instance can only be used with the exact RCollectionField that created it. Upon creation, the entry values need
+to be bound to memory locations.
 */
 // clang-format on
 class RNTupleCollectionWriter {
    friend class RCollectionField;
-   friend class RNTupleModel; // TODO(jblomer) remove me
 
 private:
-   std::size_t fBytesWritten = 0;
-   ClusterSize_t fOffset;
-   std::unique_ptr<REntry> fDefaultEntry;
+   enum class EEnvironmentState {
+      kBorn,            // created from RCollectionField
+      kConnectedToSink, // the collection field that created the writer is connected to a sink
+      kOrphaned,        // the collection field that created the writer is destructed
+   };
+
+   // fBytesWriten and fElements are reset to zero by RCollectionField::Append and thus need to be mutable
+   mutable std::size_t fBytesWritten = 0;
+   mutable ClusterSize_t fNElements;
+   /// The collection writer depends on the RCollectionField from which it was created. Filling collection elements
+   /// only works when the collection field is connected to a page sink, and as long as the collection field is alive.
+   /// This member is set by the collection field, who keeps track of all the writers it creates.
+   std::atomic<EEnvironmentState> fEnvironmentState = EEnvironmentState::kBorn;
```

Comment:
Why do we need a `std::atomic` here? Do we foresee a legitimate case of threads filling a collection writer while the model is destructed?

Reply:
The `fEnvironmentState` is edit both in the RCollectionField destructor and the RCollectionWriter destructor, and those two could end up running in different threads.

Reply:
Yes exactly, but that's still a race condition because the writer's destructor doesn't know if the field is still there or not.